### DISCUSSION TITLE
System health check API (renaming)

### DISCUSSION THREAD

**Discussion by seokho-son** [cloud-barista/cb-tumblebug#1524] [General]
# 개요
- 시스템 상태 확인을 위한 API는 서브시스템 마다 필요한 기능이며, API를 동일하게 처리하면 일관성 향상에 도움.
- 현재는 서브시스템별로 상태 확인 관련 API의 path 가 상이하거나 기능 부재. (기존 CB-TB는 GET tumblebug/`health` 로 기능 제공)
- 하기 시스템 준비 상태 확인 API 사양 제안에 따라(특별히 이견이 없는 경우), 모든 관련 서브시스템에 각자 적용
  - m-cmp 서브시스템
  - cloud-migrator 서브시스템

# 시스템 준비 상태 확인 API 사양 제안
- Path: {{subsystem_name}}/`readyz`
- Method: `GET`
- Request Body: `없음`
- Response Body: `{ "message": "your-message" }`
- HTTP status
  - API 서버가 활성화 되어 있고, 시스템의 기능 수행 준비가 완료된 상태: `200`
  - API 서버가 활성화 되어 있지만, 시스템의 기능 수행 준비가 완료되지 않은 상태: `503` (서비스 불가)

### Path 명명 사유
- 단순 API 서버 활성화 여부(health) 뿐만 아니라, 시스템 기능 수행을 위한 기본 조건 만족 여부 판단이 필요하므로, 기능을 고려하여 ready (readiness)의 의미가 포함된 명칭 활용 (향후 aliveness 확인을 지원하게 되는 경우도 고려)
- 주요 래퍼런스: Kubernetes (기존 healthz 를 deprecate 함. readyz와 livez 로 명확하게 구분)

### API 예시 (CB-Tumblebug)
https://cloud-barista.github.io/cb-tumblebug-api-web/?url=https://raw.githubusercontent.com/cloud-barista/cb-tumblebug/main/src/api/rest/docs/swagger.yaml#/%5BAdmin%5D%20System%20management/get_readyz

# 적용 방법 및 시점
- API 예시를 고려하여, API 제공
- 서브시스템별 차기 프리마이너릴리스에 반영 및 반영 현황을 각 메인테이너가 Slack 채널에 공유.

**Comment by powerkimhub** [comment]
- CB-Spider 반영 했습니다. 
  - [CB-Spider v0.8.13](https://github.com/cloud-barista/cb-spider/releases/tag/v0.8.13) 부터 제공합니다. 
  - API : [v0.8.13](https://github.com/cloud-barista/cb-spider/wiki/CB-Spider-User-Interface) / [CCTM v0.8.13](https://documenter.getpostman.com/view/24786935/2sA3JFA44u)

<br>

- Spider의 경우 다음 2가지를 체크합니다. 
  - Rest API 호출 가능 여부 확인 => `readyz` 호출 자체가 테스트
  - 메타DB 연결 이상 유무 확인 => `DB.Ping` 호출 테스트

---

#### [정상 일때]

```
curl -v http://localhost:1024/spider/readyz
```
```
*   Trying 127.0.0.1:1024...
* Connected to localhost (127.0.0.1) port 1024 (#0)
> GET /spider/readyz HTTP/1.1
> Host: localhost:1024
> User-Agent: curl/7.81.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Content-Type: application/json; charset=UTF-8
< Vary: Origin
< Date: Thu, 02 May 2024 04:06:38 GMT
< Content-Length: 33
<
{"message":"CB-Spider is ready"}
* Connection #0 to host localhost left intact
```

<br>

#### [비정상 일때:DB 호출 불가능]

```
curl -v http://localhost:1024/spider/readyz
```
```
*   Trying 127.0.0.1:1024...
* Connected to localhost (127.0.0.1) port 1024 (#0)
> GET /spider/readyz HTTP/1.1
> Host: localhost:1024
> User-Agent: curl/7.81.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 503 Service Unavailable
< Content-Type: application/json; charset=UTF-8
< Vary: Origin
< Date: Thu, 02 May 2024 04:05:40 GMT
< Content-Length: 70
<
{"message":"unable to open database file: no such file or directory"}
* Connection #0 to host localhost left intact
```

<br>

#### [비정상 일때:서버 down]
```
curl -v http://localhost:1024/spider/readyz
```
```
*   Trying 127.0.0.1:1024...
* connect to 127.0.0.1 port 1024 failed: Connection refused
*   Trying ::1:1024...
* connect to ::1 port 1024 failed: Connection refused
* Failed to connect to localhost port 1024 after 0 ms: Connection refused
* Closing connection 0
curl: (7) Failed to connect to localhost port 1024 after 0 ms: Connection refused
```

**Comment by yunkon-kim** [comment]
CM-Beetle에 반영 하였습니다.

CM-Beetle v0.1.1 부터 제공될 예정입니다.

(참고)
```bash
ubuntu@DESKTOP-L3ETGKI:~$ curl -v localhost:8056/beetle/readyz
*   Trying 127.0.0.1:8056...
* Connected to localhost (127.0.0.1) port 8056 (#0)
> GET /beetle/readyz HTTP/1.1
> Host: localhost:8056
> User-Agent: curl/7.81.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Content-Type: application/json; charset=UTF-8
< Vary: Origin
< Date: Fri, 03 May 2024 04:25:47 GMT
< Content-Length: 33
<
{"message":"CM-Beetle is ready"}
* Connection #0 to host localhost left intact
```
